portfolio/src/main/webapp/script.js

```diff
@@ -28,103 +28,233 @@ function addRandomGreeting() {
+
+
+    //Runs search on enter
     document.getElementById('searchbar').onkeydown = function(e){
         if(e.keyCode == 13){
-            runSearch(document.getElementById('searchbar')
-                .value.toLocaleLowerCase());
+            // i only want single word alphanumerics to go through
+            retrievedInput = document.getElementById('searchbar').value;
+            const regex = new RegExp(/^[a-z0-9]+$/i);
+            if (regex.test(retrievedInput) === false){
+                shakeSearchBar();
+            } else {
+                search(retrievedInput);
+            }
         }
     };
 }
 
+/**
+ * Handler function for the event listener
+ */
+function searchHandler(){
+    highlightHandler(matchingNodes, retrievedInput, match_index);
+}
+
+/**
+ * Code to shake the search bar and make it glow
+ * Calls a bunch of css transforms
+ */
+function shakeSearchBar(){
+    const el = document.getElementById('searchbar');
+    el.preventDefault;
+    el.classList.remove('searchbarglow');
+    void el.offsetWidth; // allows time for the class to be added'
+    el.classList.add('searchbarglow');
+}
+
+
 //*********************** SEARCH IMPLEMENTATION ***********************
+
+
 /**
-  * Retreives the HTML from the source
-  */
-function retrieveHtml(){
-    var source = document.body.innerHTML;
-
-    if (source != null){
-        return source.toLocaleLowerCase();    //toLocaleLowerCase() to ignore caps
-    } else {
-        alert("source is null");
+ * Finds all the occurences of a substring in a string
+ * @param {string} substr - The substring to search for.
+ * @param {string} str - The larger string to be combed through.
+ */
+function findSubstringIndices(substr, str) {
+    // no lowercase causes issues with finding the index
+    substr = substr.toLocaleLowerCase();
+    str = str.toLocaleLowerCase();
+
+    const substrLen = substr.length;
+    if (substrLen === 0) return [];
+
+    let startIndex = 0;
+    let index = 0;
+    const indices = [];
+
+    while ((index = str.indexOf(substr, startIndex)) > -1) {
+        indices.push(index);
+        startIndex = index + substrLen;
     }
+
+    return indices;
 }
 
 /**
-  * Finds the closest tag before the provided index in the html text.
-  */
-function getClosestTag(index, source){
-    while (source.charAt(index) != '>'){
-        if (index == 0){
-            alert("index hit 0");
-            break;
+ * Walks the tree and finds matches,
+ * returns index array of matches.
+ * Will optimize to gather straight from tree
+ * rather than push into new arr to save space
+ * @param {object} node - The starting point of the tree.
+ * @param {string} input - The string to be match.
+ */
+function htmlWalkerandMatcher(node, input){
+    let matches = []; 
+
+    let all = []; // queue
+    all.push(node);
+
+    while(all.length !== 0){
+        let cur = all.shift();
+
+        if (cur.nodeType==3){ // Locates text nodes. Text nodes have no child nodes. 
```

Comment:
Prefer putting comments before the line they refer to. Comments on the same line are almost exclusively for short (1-3 word) comments.